packages/library-sync/src/jvm/kotlin/io/realm/kotlin/mongodb/internal/OkHttpWebsocketClient.kt

```diff
@@ -0,0 +1,273 @@
+
+        // send any queued Frames even if the Core observer is closed, but only if the websocket is still open, this can be a message like 'unbind'
+        // which instruct the Sync server to terminate the Sync Session (server will respond by 'unbound').
+        if (!isClosed.get()) {
+            scope.launch {
+                try {
+                    if (!isClosed.get()) { // double check that the websocket is still open before sending.
+                        webSocket.send(message.toByteString())
+                        runCallback(
+                            handlerCallback,
+                            observerIsClosed.get(), // if the Core observer is closed we run this callback as cancelled (to free underlying resources)
+                            WebsocketCallbackResult.RLM_ERR_SYNC_SOCKET_SUCCESS,
+                            ""
+                        )
+                    } else {
+                        runCallback(
+                            handlerCallback,
+                            observerIsClosed.get(), // if the Core observer is closed we run this callback as cancelled (to free underlying resources)
+                            WebsocketCallbackResult.RLM_ERR_SYNC_SOCKET_CONNECTION_CLOSED,
+                            "Connection already closed"
+                        )
+                    }
+                } catch (e: Exception) {
+                    runCallback(
+                        handlerCallback,
+                        observerIsClosed.get(), // if the Core observer is closed we run this callback as cancelled (to free underlying resources)
+                        WebsocketCallbackResult.RLM_ERR_SYNC_SOCKET_RUNTIME,
+                        "Sending Frame exception: ${e.message}"
+                    )
+                }
+            }
+        } else {
+            scope.launch {
+                runCallback(
+                    handlerCallback,
+                    observerIsClosed.get(), // if the Core observer is closed we run this callback as cancelled (to free underlying resources)
+                    WebsocketCallbackResult.RLM_ERR_SYNC_SOCKET_CONNECTION_CLOSED,
+                    "Connection already closed"
+                )
+            }
+        }
+    }
+
+    override fun close() {
+        logger.debug("close")
+        observerIsClosed.set(true)
+
+        if (::webSocket.isInitialized) {
+            scope.launch {
+                if (!isClosed.get()) {
+                    webSocket.close(
+                        WebsocketErrorCode.RLM_ERR_WEBSOCKET_OK.nativeValue, "client closed websocket"
+                    )
+                }
+            }
+        }
+    }
+
+    /**
+     * Runs the [block] inside the transport [scope] only if Core didn't initiate the Websocket closure.
+     */
+    private fun runIfNotClosing(block: () -> Unit) {
```

Comment:
Nitpick: This seems to only be for blocks operating on the observer. I would have chosen a name that reflects that, because it would indicate the direction of communication and obvious not execute the block if the observer was _closed_.